### ISSUE TITLE
bug in perspective_projection

### ISSUE THREAD

**Issue by sawwhite** [eigenvivek/DiffDRR#261]
Thank you for the contribution . I noticed a strange phenomenon in the perspective_projection function of DRR. To verify this issue, I created six spheres under CT and then used the DRR algorithm for projection. When comparing the output of the DRR projection and the perspective_projection, I found significant differences when the angle changes. As shown in the figure, the bright spots represent the projection results, and the circles indicate the estimated positions from the perspective_projection.

When I use this code:
```
         rotations = torch.tensor([[ 1.0,  0., 1]], device=self.device)
         translations = torch.tensor([[ 0, 800., 0.]], device=self.device)
                        
        pred_img = drr_seg(rotations, translations, parameterization="euler_angles", convention="ZXY")
        poseIV = convert(rotations, translations, parameterization="euler_angles", convention="ZXY")
        points2d =  drr_seg.perspective_projection(poseIV,point3d)
```

![id_0_gt_bone_drr](https://github.com/eigenvivek/DiffDRR/assets/6848612/1a702aa8-2e57-4882-acd0-8d51204a1a46)

When I use this code:
```
         rotations = torch.tensor([[ .0,  0., 0]], device=self.device)
         translations = torch.tensor([[ 0, 800., 0.]], device=self.device)
                        
        pred_img = drr_seg(rotations, translations, parameterization="euler_angles", convention="ZXY")
        poseIV = convert(rotations, translations, parameterization="euler_angles", convention="ZXY")
        points2d =  drr_seg.perspective_projection(poseIV,point3d)
```
![id_0_gt_bone_drr](https://github.com/eigenvivek/DiffDRR/assets/6848612/52ca927b-76f4-4de0-8f62-9b7762ed600d)

**Comment by eigenvivek** [comment]
hi @sawwhite two questions:
- does your `DRR` module have any principal point offsets (`x0` or `y0`)? i think there's something wrong in that atm
- in the output of `drr.perspective_points`, the x and y axes are reversed from what you'd expect. does reversing this in your scatterplot fix your issue?

**Comment by sawwhite** [comment]
The issue still hasn't been resolved. Here is my code。
```
import numpy as np
import SimpleITK as sitk
import torch
from diffdrr.drr import DRR
from diffdrr.data import read
from diffdrr.data import load_example_ct,read
from diffdrr.pose import convert
import cv2

def _main():
    ct_file = '/tmp/zz_00.nii.gz'
    ct_image = sitk.ReadImage(ct_file)
    device = 'cuda'
    
    # voxel point in ct 
    point3d_voxel_xyz = torch.tensor([[[134.5000, 109.5000,  43.0000],
         [139.5000,  34.5000, 113.0000],
         [311.5000,  48.5000,  77.0000],
         [277.5000, 118.5000,  19.0000],
         [217.5000, 215.5000,  96.0000]]], device='cuda:0')

    point3d = point3d_voxel_xyz - torch.tensor(ct_image.GetSize()).unsqueeze(0).unsqueeze(0).to(device) // 2
    point3d[:,:,0] = -1 * point3d[:,:,0] # reverse x axis in ct

    seg_subject = read('/tmp/zz_00.nii.gz')
    drr_seg = DRR(
        seg_subject,
        1020.,
        512,
        1.0,
        x0= 0.,
        y0= 0.,
        reverse_x_axis=True,
        renderer  = "siddon",
    )
    drr_seg.to(device)
    rotations = torch.tensor([[1.0, 0.0, 1.0]], device=device)
    translations = torch.tensor([[0.0, 850.0, 0.0]], device=device)                        
    pred_img = drr_seg(rotations, translations, parameterization="euler_angles", convention="ZXY")

    poseIV = convert(rotations, translations, parameterization="euler_angles", convention="ZXY")
    points2d =  drr_seg.perspective_projection(poseIV,point3d)

    colors = [[0, 0, 255],[0, 255, 0],[255,0,0],[255,0,255],[255,255,0],[0,255,255]]
    out_vis = (pred_img - pred_img.min()).detach().cpu().numpy()
    out_vis = out_vis / out_vis.max()
    out_vis = (255*out_vis).astype(np.uint8)
    
    out_vis3c = cv2.cvtColor(out_vis.squeeze(), cv2.COLOR_GRAY2RGB)
    for _ii,_pp in enumerate(points2d.squeeze(0)):
        out_vis3c = cv2.circle(out_vis3c, (int(_pp[1]),int(_pp[0])), 5,  colors[_ii], 1)
    cv2.imwrite(f'/home/congc/tmp/id_gt_bone_drr101.png',out_vis3c)




if __name__ == "__main__":
    _main()
```

**Comment by sawwhite** [comment]
[zz00.nii.gz](https://github.com/user-attachments/files/15721787/zz00.nii.gz)

**Comment by coolteemf** [comment]
Hi, I'm experiencing the same bug.
First of all, @sawwhite you need take into account the fact that the affine is changed in the "read" function and convert your marker points accordingly.
You can do this by passing the markers as fiducials in the read function (you need to add a batch dimension first).
However, this part is actually problematic.
I've done as in read to convert the markers, which is to compute new_markers = new_affine @ old_affine.inv @ old_markers.
When I do that and save the subject image to nii (with image.save), and also the converted fiducials, I actually get a mismatch in 3D !
So, it looks like that, for whatever weird reason, the new_affine @ old_affine.inv matrix is not the correct frame change matrix.
![image](https://github.com/eigenvivek/DiffDRR/assets/67541941/6bebbb02-cade-41f7-aa8d-e7a1e2d53607)
As you can see, the markers (left) should be superposed with the actual markers (right).
I'll look more into it tomorrow though.

**Comment by coolteemf** [comment]
Ok, I think I got where the issue is.
The read function calls the ToCanonical function of TorchIO which calls the as_closest_canonical function of nibabel, which calls the as_reoriented method of the image.
This method does not act on the affine but rather on the data array:
https://nipy.org/nibabel/reference/nibabel.spatialimages.html#nibabel.spatialimages.SpatialImage.as_reoriented
Basically, it inverts (or not) each axis of the data array, depending on whether or not the axes match the canonical axes.

**Comment by coolteemf** [comment]
Removing the call to ToCanonical fixes the volume, but seems to mess up the projection though...
![image](https://github.com/eigenvivek/DiffDRR/assets/67541941/115cf1bb-138c-426d-990c-e9f5c3480a62)

**Comment by eigenvivek** [comment]
Hi @coolteemf thank you very much for doing some investigations into this issue.

This is a surprising result. It appears then that the `new_affine` matrix returned by `nibabel` is incorrect? This probably isn't the case since it's such a widely used package, and much more likely that I've messed up some implementation detail in the canonicalization of the fiducials. 

I have time this afternoon to do a deep dive into these functions and will see what I find. Perhaps this is also related to some of the issues with manually changing affine matrices in #252.

**Comment by coolteemf** [comment]
I think it is due to the fact that the canonicalizing func inverts some axes, which can not be done by a proper rotation, so I guess they have chosen to do it by inverting the data array directly rather than doing via an affine matrix that contains an improper rotation (ie inversion).

**Comment by coolteemf** [comment]
I implemented a fix that involves:

1. Not calling the ToCanonical() function in canonicalize
2. Replacing the use of origin and spacing to convert from and to voxel coordinates with the affine matrix of the subject inside the renderer

It works, but I don't know if this is the way you want to go about it.
If it is, I can make a PR.

**Comment by eigenvivek** [comment]
@coolteemf so are you saying that some elements of `spacing` in the renderer would be negative if the volume is in an LPS+ orientation?

**Comment by coolteemf** [comment]
@eigenvivek Yes, and if the rotation part of the affine matrix is not diagonal, it is necessary to use the affine instead of simply dividing by the spacing.

**Comment by eigenvivek** [comment]
Got it. I think your solution is reasonable and see that it would not support non-diagonal rotations in affine matrices. I would greatly appreciate a PR and will be happy to review/test it to make sure all other parts of the system are compliant.

The only thing that I know this would break is the 3D visualization functions. For example, `pyvista.ImageData` requires that the argument `spacing` is a non-negative array. So perhaps here I would would have to add some code to parse volumes inside this function.

https://github.com/eigenvivek/DiffDRR/blob/858bb1a49991abe0ad33364ce55de391d2b64090/diffdrr/visualization.py#L199-L204

Additionally, there are a few [open-source 2D/3D datasets](https://github.com/eigenvivek/DiffDRR-Datasets) for which we have ground truth camera poses. I can run your PR on these datasets and make sure we still recover the same DRR. 

If I can help with the PR in any other way, please let me know.

**Comment by eigenvivek** [comment]
Also heads up that @momen-mh and I are working on a refactor of the `diffdrr.renderers` module (#272). It only changes how voxels are looked up in Siddon's method, so it shouldn't affect your PR. It should be merged by this evening.
